### PR TITLE
Create empty entries for nonexistent subtrees in cache preload

### LINKED ISSUE #345: "Unexpectedly reading from within GetNodeHash()" warnings
If I add `--alsologtostderr` to the invocation of `./trillian_map_server` in `integration/map_integration_test.sh`, I see lots of warnings:

```
W0131 09:37:43.527794   11656 subtree_cache.go:185] Unexpectedly reading from within GetNodeHash()
```
The comment there says "This should never happen - we should've already read all the data we need above, in Preload()"...

### PR DESCRIPTION
**Pull request by Martin2112**
Create empty subtrees in cache for any that preload fails to retrieve because they don't exist yet. This prevents a double fetch of these subtrees, which will be unsuccessful both times and logged warnings for each time this would happen. This was common for maps because of the way the sparse merkle tree works but not an issue for log trees.

Closes #345.

### CODE REVIEW

File: storage/cache/subtree_cache.go [hunk 1]
```diff
@@ -196,7 +201,25 @@ func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc
...
 	}
+
+	// We might not have got all the subtrees we requested, if they don't already exist.
+	// Create empty subtrees for anything left over. As an additional sanity check we refuse
+	// to overwrite anything already in the cache as we determined above that these subtrees
+	// should not exist in the subtree cache map.
+	for _, id := range want {
+		prefixLen := id.PrefixLenBits / depthQuantum
+		px := id.Path[:prefixLen]
+		t := s.newEmptySubtree(*id, px)
```
Comment (AlCutter) [inline storage/cache/subtree_cache.go]: Move this creation down to L217 where it's used (handily avoids creating it if we don't need to)?

File: storage/cache/subtree_cache.go [hunk 2]
```diff
@@ -196,7 +201,25 @@ func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc
...
+		_, ok := s.subtrees[string(px)]
+		if !ok {
```
Comment (AlCutter) [inline storage/cache/subtree_cache.go]: Should probably reverse the sense of the test here:
`go
if ok {
  // it's actually not ok - indent error flow
  glog...
  continue
}`

and outdent the good case (could rename `ok` to `found` or something to reduce confusion that `ok==true` is the "bad" case)

File: storage/cache/subtree_cache.go [hunk 3]
```diff
@@ -196,7 +201,25 @@ func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc
...
+			s.subtrees[string(px)] = t
+		} else {
+			glog.Warningf("Preload tried to clobber existing subtree for: %v", *id)
```
Comment (AlCutter) [inline storage/cache/subtree_cache.go]: This is probably worth an error return, something's gone quite wrong to end up with this situation...

File: storage/cache/subtree_cache.go (hunk 1 above)
Comment (Martin2112) [inline storage/cache/subtree_cache.go]: Done.

File: storage/cache/subtree_cache.go (hunk 2 above)
Comment (Martin2112) [inline storage/cache/subtree_cache.go]: Don't need the ugliness of continue thankfully as reversing the test + returning an error removes the need for it.

File: storage/cache/subtree_cache.go (hunk 3 above)
Comment (Martin2112) [inline storage/cache/subtree_cache.go]: Yeah, converted to an error.

File: storage/cache/subtree_cache.go [hunk 4]
```diff
@@ -211,13 +211,11 @@ func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc
 	for _, id := range want {
 		prefixLen := id.PrefixLenBits / depthQuantum
 		px := id.Path[:prefixLen]
-		t := s.newEmptySubtree(*id, px)
-		_, ok := s.subtrees[string(px)]
-		if !ok {
-			s.subtrees[string(px)] = t
-		} else {
-			glog.Warningf("Preload tried to clobber existing subtree for: %v", *id)
+		_, exists := s.subtrees[string(px)]
```
Comment (AlCutter) [inline storage/cache/subtree_cache.go]: factor out `string(px)` since it's used below too?

Comment (Martin2112) [inline storage/cache/subtree_cache.go]: Done.

**Pull request merged**